components/style/properties/helpers/animated_properties.mako.rs

```diff
@@ -883,3 +883,68 @@ impl ToAnimatedZero for AnimatedFilter {
         }
     }
 }
+
+/// An iterator over all the properties that transition on a given style.
+pub struct TransitionPropertyIterator<'a> {
+    style: &'a ComputedValues,
+    index_range: core::ops::Range<usize>,
+    longhand_iterator: Option<NonCustomPropertyIterator<LonghandId>>,
+}
+
+impl<'a> TransitionPropertyIterator<'a> {
+    /// Create a `TransitionPropertyIterator` for the given style.
+    pub fn from_style(style: &'a ComputedValues) -> Self {
+        Self {
+            style,
+            index_range: 0..style.get_box().transition_property_count(),
+            longhand_iterator: None,
+        }
+    }
+}
+
+/// A single iteration of the TransitionPropertyIterator.
+pub struct TransitionPropertyIteration {
+    /// The id of the longhand for this property.
+    pub longhand_id: LonghandId,
+
+    /// The index of this property in the list of transition properties for this
+    /// iterator's style.
+    pub index: usize,
+}
+
+impl<'a> Iterator for TransitionPropertyIterator<'a> {
+    type Item = TransitionPropertyIteration;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        use crate::values::computed::TransitionProperty;
+        loop {
+            if let Some(longhand_id) = self.longhand_iterator
+                .as_mut()
+                .and_then(|iterator| iterator.next())
+            {
+                return Some(TransitionPropertyIteration {
+                    longhand_id,
+                    index: self.index_range.start,
+                });
+            }
+
+            let index = self.index_range.next()?;
+            match self.style.get_box().transition_property_at(index) {
+                TransitionProperty::Longhand(longhand_id) => {
+                    self.longhand_iterator = None;
```

Comment:
Would maybe be a bit cleaner to do:

```
if let Some(ref mut it) = self.longhand_iterator {
    if let Some(longhand_id) = it.next() {
        return Some(...);
    }
    self.longhand_iterator = None;
}
```

I think that should work with nll but otherwise you can move the `= None` assignment after the `if`. That should allow you to remove this assignment and the custom / unsupported property one (and it's also nicer to clean up our state asap). Wdyt?

Reply:
Yep. All this makes sense. I've pushed a new version of the branch that makes these changes.